src/amber/cli/recipes/recipe_fetcher.cr

```diff
@@ -0,0 +1,72 @@
+require "http/client"
+require "zip"
+
+module Amber::Recipes
+  class RecipeFetcher
+    getter kind : String # one of the supported kinds [app, controller, scaffold]
+    getter name : String | Nil
+    getter directory : String
+
+    def initialize(@kind : String, @name : String | Nil)
+      @directory = "#{Dir.current}/#{@name}/#{@kind}"
+    end
+
+    def fetch
+      if Dir.exists?(@directory)
+        return @directory
+      end
+
+      if Dir.exists?("#{@name}/#{@kind}")
+        return "#{@name}/#{@kind}"
+      end
+
+      template_path = "#{AMBER_RECIPE_FOLDER}/#{@name}"
+
+      if Dir.exists?("#{template_path}/#{@kind}")
+        return "#{template_path}/#{@kind}"
+      end
+
+      if Dir.exists?("../#{template_path}/#{@kind}")
+        return "../#{template_path}/#{@kind}"
+      end
+
+      return fetch_url template_path
+    end
+
+    def recipe_source
+      CLI.config.recipe_source || "https://raw.githubusercontent.com/amberframework/recipes/master/"
```

Comment:
Recipes should be shards not zip files.  This is adding another packaging system to the mix with the need to handle versioning, compatibility, dependencies, etc.

Reply:
Distributing recipes as shards feels like it'll be quite a wedge. They're not versioned, have no dependencies, and really aren't even crystal code. They're just collections of templates and folders. A zip is a great way to package a recipe for download, I just don't think that this way of organizing and distributing "official" amber recipes is going to scale well.

If Amber is really going to provide a handful of blessed recipes, I see no reason why they can't all live in the same repository. But I'd like them to live _as code_, not as zips, so pull requests are functional and file history works. A CI system can create zips and publish them to s3 or something for distribution. 

I think it also makes sense for others to be able to publish a recipe as a single git repository, but keeping up with N repositories for blessed recipes feels exausting. Amber is already 21 separate repositories. Imagine trying to update for a breaking change across N repositories, that requires coordinating pull requests and branches across every repository to match the amber core.

Reply:
@robacarp Nice! so you mean, we should maintain just source code in the `amberframework/recipes` repo and then generate zips using CI :+1: 

@damianham WDYT?

Reply:
I don't see a need for multiple repositories, the recipes can be organised in sub folders from the root 
amberframework/recipes/react
amberframework/recipes/vue
amberframework/recipes/angular
amberframework/recipes/plugins

a recipe name can include path separators e.g.  -r react/modular/streaming to use http://amberframework.org/recipes/react/modular/streaming.zip
if the recipes are published to zips on the website by CI

Reply:
hey, sorry I'm just curious why not just use git and pull the repo? It automaticly handles verification you can pull from public or private repo, and don't need the step with creating/updating/managing zip file

Reply:
@Adam-Stomski That will need to clone the entire repo before using it, so if the repo grows enough, we will be unable to use recipes because is so big.

I think keeping recipes source code is fine, then create zip files from source and storing them inside recipe directory is a nice option.

Reply:
This thread is getting big :sweat_smile: 

I opened a tracking issue for this, please leave your comments here: https://github.com/amberframework/amber/issues/750